File: lib/domain/usecases/task/create_task_usecase.dart
import '../../entities/commit.dart' as entity;
import '../../entities/enums.dart';
import '../../entities/task.dart';
import '../../repositories/i_commit_repository.dart';
import '../../repositories/i_task_repository.dart';

/// 创建任务用例
class CreateTaskUseCase {
  CreateTaskUseCase(
    this._taskRepository,
    this._commitRepository,
  );

  final ITaskRepository _taskRepository;
  final ICommitRepository _commitRepository;

  Future<Task> execute({
    required String branchId,
    required String title,
    String? description,
    Priority priority = Priority.medium,
    DateTime? dueDate,
    String? parentTaskId,
  }) async {
    if (title.trim().isEmpty) {
      throw ArgumentError('任务标题不能为空');
    }

    final task = Task.create(
      branchId: branchId,
      title: title.trim(),
      description: description,
      priority: priority,
      dueDate: dueDate,
      parentTaskId: parentTaskId,
    );

    final createdTask = await _taskRepository.create(task);

    // 记录创建提交
    await _commitRepository.create(
      entity.Commit.create(
        taskId: createdTask.id,
        branchId: branchId,
        message: '创建任务: ${createdTask.title}',
        type: CommitType.create,
      ),
    );

    return createdTask;
  }
}

File: web/src/application/usecases/task/create-task-usecase.ts
import { Commit } from '../../../domain/entities/commit';
import { CommitType, Priority } from '../../../domain/entities/enums';
import { Task } from '../../../domain/entities/task';
import { ICommitRepository } from '../../../domain/repositories/i-commit-repository';
import { ITaskRepository } from '../../../domain/repositories/i-task-repository';
import { validateNotEmpty } from '../../../core/utils/validators';

export interface CreateTaskInput {
  branchId: string;
  title: string;
  description?: string | null;
  priority?: Priority;
  dueDate?: Date | null;
  parentTaskId?: string | null;
}

export class CreateTaskUseCase {
  constructor(
    private taskRepo: ITaskRepository,
    private commitRepo: ICommitRepository
  ) {}

  async execute(input: CreateTaskInput): Promise<Task> {
    validateNotEmpty(input.branchId, '分支 ID');
    validateNotEmpty(input.title, '任务标题');
    const task = Task.create(input.branchId, input.title, {
      description: input.description,
      priority: input.priority,
      dueDate: input.dueDate,
      parentTaskId: input.parentTaskId,
    });
    const created = await this.taskRepo.create(task);
    const commit = Commit.create(
      created.id,
      created.branchId,
      `创建任务 ${created.title}`,
      CommitType.create
    );
    await this.commitRepo.create(commit);
    return created;
  }
}
